### ISSUE TITLE
Dropout Bug in Prenet

### ISSUE THREAD

**Issue by npuichigo** [keithito/tacotron#165]
I find a huge bug about dropout in Prenet, and maybe it's the final reason that your repo's implementation often cannot get good alignments with small reduction factor.

In tf.layers.dropout, you don't set the training flag to true when training. So in dropout's implementation, it just applies an identity op on the inputs.

![image](https://user-images.githubusercontent.com/11533479/39663460-6561e6b2-50a6-11e8-8f3c-4f0d499bf207.png)

<img width="751" alt="image" src="https://user-images.githubusercontent.com/11533479/39663507-42c3b44a-50a7-11e8-8bb8-98e4fdb4b64e.png">

<img width="815" alt="image" src="https://user-images.githubusercontent.com/11533479/39663508-5220ce0a-50a7-11e8-9f49-1419e29fed18.png">

**Comment by keithito** [comment]
Oh wow, I think you're right. That is a huge bug! I'll try training with that fixed and see what happens.

**Comment by yishaicohe** [comment]
hey,
any conclusions so far?

**Comment by lautjy** [comment]
Nice catch!  Also interested in this.
Going to do some augmentation tests soon, so would be nice to have insight if this was really missing previously or not.

**Comment by keithito** [comment]
After making the change to fix dropout, the model eventually learns to align during training, but there seem to be alignment issues more frequently during eval, and subjectively, the output sounds worse. This is the opposite of what I would have expected, and I need to look into why -- but unfortunately, I don't have a whole lot of time these days...

**Comment by npuichigo** [comment]
Can you now get alignment using small reduction factor? What about using dropout during eval to introduce output variance?

**Comment by bfs18** [comment]
Hi @keithito  @npuichigo , it seems that `training=True, droprate=0.0` is equivalent to `training=False`. It is wired that the results are different.
nn.dropout returns `x / (1 - rate) * dropout_mask`.
layer.Dropout returns `nn.dropout(x, 1 - rate) if training else x` 
So if `training=False`, layer.Dropout returns `x`, if `training=True and rate=0.0`, layer.Dropout also returns `x`, since dropout_mask are all `1`s.

**Comment by npuichigo** [comment]
Hi @keithito, you said that when fixing dropout, the model still cannot learn to align during training before. I found that there maybe another mistake that you should put DecoderPrenetWrapper outside AttentionWrapper. Otherwise, because the AttentionWrapper concatenates the input and previous attention vector before calling the internal cell, dropout in the DecoderPrenetWrapper will drop 50% information of previous attention vector, which makes alignments hard to learn.

Here's what I mean.
```python
attention_cell = AttentionWrapper(
        GRUCell(hp.attention_depth),
        BahdanauAttention(hp.attention_depth, encoder_outputs),
        alignment_history=True,
        output_attention=False)
attention_cell = DecoderPrenetWrapper(attention_cell, is_training, hp.prenet_depths)
```
